File: src/main/java/com/koushikdutta/scratch/buffers/ReadableBuffers.java
package com.koushikdutta.scratch.buffers;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.Charset;

public interface ReadableBuffers {
    ByteOrder order();
    void order(ByteOrder order);

    default boolean isEmpty() {
        return remaining() == 0;
    }
    default boolean hasRemaining() {
        return !isEmpty();
    }
    int remaining();

    /**
     * Skip the given number of bytes.
     * @param length
     * @return
     */
    ReadableBuffers skip(int length);

    /**
     * Empty the buffer.
     */
    void free();

    default byte[] getBytes(int length) {
        byte[] ret = new byte[length];
        get(ret);
        return ret;
    }
    default byte[] getBytes() {
        return getBytes(remaining());
    }
    byte[] peekBytes(int size);
    ByteBuffer[] getAll();

    void get(byte[] bytes);
    void get(byte[] bytes, int offset, int length);
    void get(WritableBuffers into, int length);

    /**
     * Read the buffer. Returns false if nothing was read due to the buffer being empty.
     * @param into
     * @return
     */
    boolean get(WritableBuffers into);

    /**
     * Fill the given buffer.
     * @param buffer
     */
    void get(ByteBuffer buffer);
    /**
     * Return all available data as a single buffer.
     * @return
     */
    ByteBuffer getByteBuffer();
    ByteBuffer getByteBuffer(int length);
    int getInt();
    char getByteChar();
    short getShort();
    byte get();
    long getLong();
    default String getString() {
        return getString(null);
    }
    default String getString(int length, Charset charset) {
        return new String(this.getBytes(length), charset);
    }
    default String getString(Charset charset) {
        String ret = peekString(charset);
        free();
        return ret;
    }

    /**
     * Fill the given buffer with data, until the given sequence of bytes is found.
     * @param into
     * @param scan The byte sequence to find.
     * @return Returns true if the byte sequence was found, false if the buffer ends
     * before the sequence was found.
     * For example, if the this buffer ends on a partial byte sequence match, the partial sequence
     * will be left in the this buffer, and all data prior to that sequence will be filled into
     * the given buffer.
     */
    boolean getScan(WritableBuffers into, byte[] scan);

    default void spewString() {
        System.out.println(peekString());
    }

    char peekByteChar();
    short peekShort();
    byte peek();
    int peekInt();
    long peekLong();
    default String peekString() {
        return peekString(null);
    }
    String peekString(Charset charset);
}

File: buffer/src/commonMain/kotlin/com/koushikdutta/scratch/buffers/ReadableBuffers.kt
package com.koushikdutta.scratch.buffers

interface ReadableBuffers : AllocatingBuffers {
    val isEmpty: Boolean
        get() = remaining() == 0

    fun order(): ByteOrder
    fun order(order: ByteOrder)
    fun hasRemaining(): Boolean {
        return !isEmpty
    }

    fun remaining(): Int

    /**
     * Skip the given number of bytes.
     * @param length
     * @return
     */
    fun skip(length: Int): ReadableBuffers

    /**
     * Empty the buffer.
     */
    fun free()

    fun readBytes(length: Int = remaining()): ByteArray {
        val ret = ByteArray(length)
        read(ret)
        return ret
    }

    fun peekBytes(size: Int): ByteArray
    fun readAll(): Array<ByteBuffer>

    fun read(bytes: ByteArray)
    fun read(bytes: ByteArray, offset: Int, length: Int)
    fun read(into: WritableBuffers, length: Int)

    /**
     * Read the buffer. Returns false if nothing was read due to the buffer being empty.
     * @param into
     * @return
     */
    fun read(into: WritableBuffers): Boolean

    /**
     * Fill the given buffer.
     * @param buffer
     */
    fun read(buffer: ByteBuffer)

    /**
     * Return all available data as a single buffer.
     * @return
     */
    fun readByteBuffer(): ByteBuffer

    /**
     * Return all available data as a single direct buffer.
     * @return
     */
    fun readDirectByteBuffer(): ByteBuffer

    /**
     * Read the first ByteBuffer
     */
    fun readFirst(): ByteBuffer
    fun readByteBuffer(length: Int): ByteBuffer
    fun readInt(): Int
    fun readByteChar(): Char
    fun readShort(): Short
    fun readByte(): Byte
    fun readLong(): Long
    fun readUtf8String(length: Int): String {
        return this.readBytes(length).decodeToString()
    }

    fun readUtf8String(): String {
        val ret = peekUtf8String()
        free()
        return ret
    }

    /**
     * Fill the given buffer with data, until the given sequence of bytes is found.
     * @param into
     * @param scan The byte sequence to find.
     * @return Returns true if the byte sequence was found, false if the buffer ends
     * before the sequence was found.
     * For example, if the this buffer ends on a partial byte sequence match, the partial sequence
     * will be left in the this buffer, and all data prior to that sequence will be filled into
     * the given buffer.
     */
    fun readScan(into: WritableBuffers, scan: ByteArray): Boolean

    open fun spewString() {
        println(this.peekUtf8String())
    }

    fun peekByteChar(): Char
    fun peekShort(): Short
    fun peek(): Byte
    fun peekInt(): Int
    fun peekLong(): Long
    fun peekUtf8String(): String
}
